README.md
# D3 Latency Heatmap

This is a reusable D3 latency heatmap chart, which is a highly
effective way to visualize latency data over time.  The chart is
fast, easy to use, and produces beautiful images such as the
below:

![Example Latency Heatmap](samples/report-queue-latency.png)

For more on latency heatmaps, see:
* [Brendan Gregg's Latency Heat Maps](http://www.brendangregg.com/HeatMaps/latency.html)
* My blog post series on [Visualizing Latency](https://stevenengelhardt.com/post-series/visualizing-latency-2017/)
* [Visualizing System Latency](https://queue.acm.org/detail.cfm?id=1809426)

## Installing

If you use NPM, `npm install d3-latency-heatmap`.  Otherwise, download
the [latest release binaries](https://sengelha.github.io/d3-latency-heatmap/d3-latencyHeatmap.v1.zip)
or the [latest release source code](https://github.com/sengelha/d3-latency-heatmap/releases/latest).
You can also load directly from
[unpkg.com](https://unpkg.com/d3-latency-heatmap/).

For convenience, you can use the [latest release
binaries](https://sengelha.github.io/d3-latency-heatmap/d3-latencyHeatmap.v1.zip)
with D3 version 4.x as follows (D3 version 3.x is not supported):

```html
<script src="//d3js.org/d3.v4.min.js"></script>
<script src="//sengelha.github.io/d3-latency-heatmap/d3-latencyHeatmap.v1.min.js"></script>

<div id="chart"></div>
<script>
(function() {
    var parseTime = d3.timeParse("%Y-%m");
    var chart = d3.latencyHeatmap()
        .x(function (d) { return parseTime(d.date); })
        .y(function (d) { return +d.bucket; })
        .yFormat(function(d) { return d + " s"; })
        .count(function(d) { return +d.count; })
        .colorRange([d3.rgb('#FFFFFF'), d3.rgb('#5B82A1')])
        .tooltipText(function (d) { return "YearMonth: " + d[0].toISOString().substring(0, 7) + "\nBucket: " + d[1] + "\nCount: " + d[2]; })
        .rectSize([8, 8]);
    d3.csv("//sengelha.github.io/d3-latency-heatmap/sample-data.csv", function (data) {
        var svg = d3.select("#chart")
            .datum(data)
            .call(chart);
    });
})();
</script>
```

## Usage Notes

* In order to use this chart, your data must already be bucketized
  -- you cannot use raw events.  A bucket contains three data points:
  the x-value, the y-value, and the number of observations in this
  bucket.
* The color intensity of a cell is determined based on its value
  relative to the maximum value of all buckets.  This implies that
  if the number of observations per x-value (e.g. per day) increases
  over time, the cells will start with faint colors and grow more
  intense over time.
* This chart's API was modelled upon Mike Bostock's
  [Toward Reusable Charts](https://bost.ocks.org/mike/chart/) proposal.

## Examples

* [Report Queue Latency](samples/report-queue-latency.html)
* [Animated Heatmap](samples/animated-heatmap.html)
* [Using with Unpkg](samples/unpkg-modules.html)

## API Reference

### d3.latencyHeatmap()

Creates a new latency heatmap chart which may later be rendered into a
container.  Returns a *latencyHeatmap* object.

The typical pattern that the chart is rendered is by:
1. Selecting the container (e.g. a `div`) into which the chart will be rendered
2. Assigning the data to the container
3. Using d3's `call()` method.

Example:
```javascript
var data = [...]; // May be sourced using d3.csv(), d3.json(), etc.
d3.select("#container")
    .datum(data)
    .call(chart);
```

### *latencyHeatmap*.colorRange(*[minColor, maxColor]*)

Defines the color range to be used when filling cells.  This color
range will be interpolated using `d3.interpolateRgb()`.  If not set,
defaults to `[d3.rgb('#FFFFFF'), d3.rgb('#F03524')]`.

Example:
```javascript
d3.latencyHeatmap()
    .colorRange([d3.rgb('#FFFFFF'), d3.rgb('#5B82A1')]);
```

### *latencyHeatmap*.count(*accessor*)

Defines an count accessor which is called for each row in `data`.
Must return a number, which corresponds to the number of observations
within the bucket.  If not set, defaults to `function(d) { return d[2]; }`.

Example:
```javascript
d3.latencyHeatmap()
    .count(function(d) { return d.count; }); // d is { x: Date, y: number, count: number }
```

### *latencyHeatmap*.height(*h*)

Sets the height of the rendered chart to *h*.  Automatically scales
the size of the drawn rectangles to fit the specified chart height.
If not set, defaults to 400.

This value is ignored if the rectangle size is set using
**rectSize**().

Example:
```javascript
d3.latencyHeatmap()
    .height(400);
```

### *latencyHeatmap*.rectSize(*[w, h]*)

Sets the size of the individual rectangles used to draw the chart to
be width *w* and height *w*.  When set, the chart automatically
calculates the total width and height based on the number of elements
to be drawn.

Example:
```javascript
d3.latencyHeatmap()
    .rectSize([6, 4]);
```

### *latencyHeatmap*.tooltipText(*formatter*)

Defines an accessor which can be used to control how tooltips for
each drawn rectangle are formatted.  *formatter* is called with an
array with three elements: the x-value for the tick (a `Date` object),
the y-value for the tick (a number) and the count.  If not set, no
tooltips are drawn.

Example:
```javascript
d3.latencyHeatmap()
    .tooltipText(function (d) { return "X: " + d[0] + " Y: " + d[1] + " Count: " + d[2]; });
```

### *latencyHeatmap*.width(*w*)

Sets the width of the rendered chart to *w*.  Automatically scales
the size of the drawn rectangles to fit the specified chart width.
If not set, defaults to 600.

This value is ignored if the rectangle size is set using
**rectSize**().

Example:
```javascript
d3.latencyHeatmap()
    .width(600);
```

### *latencyHeatmap*.x(*accessor*)

Defines an x accessor which is called for each row in `data`.  Must return
a `Date` object, which must correspond to the timestmap of the bucket.
If not set, defaults to `function(d) { return d[0]; }`.

Example:
```javascript
d3.latencyHeatmap()
    .x(function(d) { return d.x; }); // d is { x: Date, y: number, count: number }
```

### *latencyHeatmap*.xFormat(*formatter*)

Defines an accessor which can be used to control how tick labels on the
x-axis are formatted.  *formatter* is called with the x-value for the tick,
which is a `Date` object.  If not set, defaults to the d3 default time
axis tick formatter.

Example:
```javascript
d3.latencyHeatmap()
    .xFormat(function(dt) { return dt.toLocaleString(); });
```

### *latencyHeatmap*.y(*accessor*)

Defines an y accessor which is called for each row in `data`.  Must return
a number, which corresponds to the y-value of the bucket.
If not set, defaults to `function(d) { return d[1]; }`.

Example:
```javascript
d3.latencyHeatmap()
    .y(function(d) { return d.y; }); // d is { x: Date, y: number, count: number }
```

### *latencyHeatmap*.yFormat(*accessor*)

Defines an accessor which can be used to control how tick labels on the
y-axis are formatted.  *formatter* is called with the y-value for the tick.
If not set, defaults to the d3 default linear axis tick formatter.

Example:
```javascript
d3.latencyHeatmap()
    .yFormat(function(y) { return y + " ms"; }); // y values are denoted in ms
```

## License

This project is licensed under the MIT License.


CHANGELOG.md
# 1.0 - 2017-11-15
Initial release.  Supports custom width, height, rectSize, color ranges,
and formattable axis labels.

# 1.1 - 2017-12-05
- Added support for tooltips via `tooltipText()` method

# 1.2 - 2018-01-09
- Rename version 1 of latency heatmap to .v1.js to prepare for version 2

test/latencyHeatmap-test.js
var tape = require("tape"),
    d3 = require("../"),
    d3c = require("d3-color");

tape("latencyHeatmap() initialized with expected defaults per documentation", function(test) {
  var chart = d3.latencyHeatmap();
  test.deepEqual(chart.colorRange(), [d3c.rgb('#FFFFFF'), d3c.rgb('#F03524')]);
  test.equal(chart.height(), 400);
  test.equal(chart.width(), 600);
  test.end();
});

latency-heatmap.js
(function () {
    d3.latencyHeatmap = function () {
        var margin = { top: 1, right: 1, bottom: 11, left: 1 },
            width = 600,
            height = 400,
            rectSize = undefined,
            xAccessor = function (d) { return d[0]; },
            xFormat = undefined,
            yAccessor = function (d) { return d[1]; },
            yFormat = undefined,
            countAccessor = function (d) { return d[2]; },
            colorRange = [d3.rgb('#FFFFFF'), d3.rgb('#F03524')],
            axisColor = "#ddd",
            xAxisTextColor = "#999",
            yAxisTextColor = "#000",
            axisFontSizePx = 11;

        function chart(selection) {
            selection.each(function (data) {
                // Convert data to standard representation
                data = data.map(function (d, i) {
                    return [xAccessor.call(data, d, i), yAccessor.call(data, d, i), countAccessor.call(data, d, i)];
                });

                // If rectSize is undefined, its dynamically calculated based
                // on width & height.  Otherwise, width & height are dynamically
                // calculated based on rectSize.
                var nCols = d3.nest()
                    .key(function (d) { return d[0] })
                    .entries(data)
                    .length;
                var nRows = d3.nest()
                    .key(function (d) { return d[1] })
                    .entries(data)
                    .length;
                if (rectSize) {
                    width = rectSize[0] * nCols + margin.left + margin.right;
                    height = rectSize[1] * nRows + margin.top + margin.bottom;
                } else {
                    rectSize = [(width - margin.left - margin.right) / nCols, (height - margin.top - margin.bottom) / nRows];
                }

                var contentWidth = width - margin.left - margin.right,
                    contentHeight = height - margin.top - margin.bottom;

                // Calculate all the extents once
                var xExtent = d3.extent(data, function (d) { return d[0]; });
                var yExtent = d3.extent(data, function (d) { return d[1]; });
                var countExtent = d3.extent(data, function (d) { return d[2]; });

                var xScale = d3.time.scale()
                    .domain(xExtent)
                    .range([0, contentWidth - rectSize[0]]);
                var yAxisScale = d3.scale.linear()
                    .domain(yExtent)
                    .range([contentHeight, 0]);
                var yContentScale = d3.scale.linear()
                    .domain(yExtent)
                    .range([contentHeight - rectSize[1], 0]);
                var countScale = d3.scale.linear()
                    .domain(d3.extent(data, function (d) { return d[2]; }))
                    .interpolate(d3.interpolateRgb)
                    .range(colorRange);

                // Create the axis definitions using the scale
                var xAxisDef = d3.svg.axis()
                    .scale(xScale)
                    .orient("bottom")
                    .innerTickSize(-contentHeight)
                    .outerTickSize(0);
                if (xFormat)
                    xAxisDef.tickFormat(xFormat);
                var yAxisDef = d3.svg.axis()
                    .scale(yAxisScale)
                    .orient("left")
                    .innerTickSize(-contentWidth)
                    .outerTickSize(0)
                    .ticks(1);
                if (yFormat)
                    yAxisDef.tickFormat(yFormat);

                // Select the svg element, if it exists.
                var svg = d3.select(this).selectAll("svg").data([data]);
                // Otherwise, create the skeletal chart.
                var gEnter = svg.enter().append("svg").append("g");
                gEnter.append("g").attr("class", "elems");
                gEnter.append("g").attr("class", "x axis");
                gEnter.append("g").attr("class", "y axis");
                
                // Update the outer dimensions.
                svg.attr("width", width)
                    .attr("height", height);

                // Update the inner dimensions.
                var g = svg.select("g")
                    .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

                // Update the individual points
                var nodes = g.select(".elems")
                    .selectAll("rect")
                    .data(data);
                nodes.enter()
                    .append("rect")
                    .attr("x", function (d) { return xScale(d[0]); })
                    .attr("y", function (d) { return yContentScale(d[1]); })
                    .attr("width", rectSize[0])
                    .attr("height", rectSize[1])
                    .attr("fill", function (d) { return countScale(d[2]); });
                nodes.exit()
                    .remove();

                // Update the x-axis
                var xAxis = g.select(".x.axis")
                    .attr("transform", "translate(0," + contentHeight + ")")
                    .style("font-size", axisFontSizePx + "px")
                    .call(xAxisDef);
                xAxis.selectAll("line")
                    .style("shape-rendering", "crispEdges")
                    .style("stroke", axisColor);
                xAxis.selectAll("text")
                    .style("text-anchor", "start")
                    .style("fill", xAxisTextColor);

                // Update the y-axis
                var yAxis = g.select(".y.axis")
                    .style("font-size", axisFontSizePx + "px")
                    .call(yAxisDef);
                yAxis.selectAll("line")
                    .style("shape-rendering", "crispEdges")
                    .style("stroke", axisColor);
                yAxis.selectAll(".tick text")
                    .attr("x", 2)
                    .style("text-anchor", "start")
                    .style("fill", yAxisTextColor);
                yAxis.selectAll(".tick:last-of-type text")
                    .attr("y", 8);
                yAxis.selectAll(".tick:first-of-type text")
                    .attr("y", -8);
            });
        }

        chart.width = function (_) {
            if (!arguments.length) return width;
            width = _;
            return chart;
        }

        chart.height = function (_) {
            if (!arguments.length) return height;
            height = _;
            return chart;
        }

        chart.rectSize = function (_) {
            if (!arguments.length) return rectSize;
            rectSize = _;
            return chart;
        }

        chart.x = function (_) {
            if (!arguments.length) return xAccessor;
            xAccessor = _;
            return chart;
        }

        chart.y = function (_) {
            if (!arguments.length) return yAccessor;
            yAccessor = _;
            return chart;
        }

        chart.count = function (_) {
            if (!arguments.length) return countAccessor;
            countAccessor = _;
            return chart;
        }

        chart.xFormat = function (_) {
            if (!arguments.length) return xFormat;
            xFormat = _;
            return chart;
        }

        chart.yFormat = function (_) {
            if (!arguments.length) return yFormat;
            yFormat = _;
            return chart;
        }

        chart.colorRange = function (_) {
            if (!arguments.length) return colorRange;
            colorRange = _;
            return chart;
        }

        return chart;
    };
})();

latency-heatmap.v3.js
(function () {
    d3.latencyHeatmap = function () {
        var margin = { top: 1, right: 1, bottom: 11, left: 1 },
            width = 600,
            height = 400,
            rectSize = undefined,
            xAccessor = function (d) { return d[0]; },
            xFormat = undefined,
            yAccessor = function (d) { return d[1]; },
            yFormat = undefined,
            countAccessor = function (d) { return d[2]; },
            tooltipText = undefined,
            colorRange = [d3.rgb('#FFFFFF'), d3.rgb('#F03524')],
            axisColor = "#ddd",
            xAxisTextColor = "#999",
            yAxisTextColor = "#000",
            axisFontSizePx = 11;

        function chart(selection) {
            selection.each(function (data) {
                // Convert data to standard representation
                data = data.map(function (d, i) {
                    return [xAccessor.call(data, d, i), yAccessor.call(data, d, i), countAccessor.call(data, d, i)];
                });

                // If rectSize is undefined, its dynamically calculated based
                // on width & height.  Otherwise, width & height are dynamically
                // calculated based on rectSize.
                var nCols = d3.nest()
                    .key(function (d) { return d[0] })
                    .entries(data)
                    .length;
                var nRows = d3.nest()
                    .key(function (d) { return d[1] })
                    .entries(data)
                    .length;
                if (rectSize) {
                    width = rectSize[0] * nCols + margin.left + margin.right;
                    height = rectSize[1] * nRows + margin.top + margin.bottom;
                } else {
                    rectSize = [(width - margin.left - margin.right) / nCols, (height - margin.top - margin.bottom) / nRows];
                }

                var contentWidth = width - margin.left - margin.right,
                    contentHeight = height - margin.top - margin.bottom;

                // Calculate all the extents once
                var xExtent = d3.extent(data, function (d) { return d[0]; });
                var yExtent = d3.extent(data, function (d) { return d[1]; });
                var countExtent = d3.extent(data, function (d) { return d[2]; });

                var xScale = d3.time.scale()
                    .domain(xExtent)
                    .range([0, contentWidth - rectSize[0]]);
                var yAxisScale = d3.scale.linear()
                    .domain(yExtent)
                    .range([contentHeight, 0]);
                var yContentScale = d3.scale.linear()
                    .domain(yExtent)
                    .range([contentHeight - rectSize[1], 0]);
                var countScale = d3.scale.linear()
                    .domain(countExtent)
                    .interpolate(d3.interpolateRgb)
                    .range(colorRange);

                // Create the axis definitions using the scale
                var xAxisDef = d3.svg.axis()
                    .scale(xScale)
                    .orient("bottom")
                    .innerTickSize(-contentHeight)
                    .outerTickSize(0);
                if (xFormat)
                    xAxisDef.tickFormat(xFormat);
                var yAxisDef = d3.svg.axis()
                    .scale(yAxisScale)
                    .orient("left")
                    .innerTickSize(-contentWidth)
                    .outerTickSize(0)
                    .ticks(1);
                if (yFormat)
                    yAxisDef.tickFormat(yFormat);

                // Select the svg element, if it exists.
                var svg = d3.select(this).selectAll("svg").data([data]);
                // Otherwise, create the skeletal chart.
                var gEnter = svg.enter().append("svg").append("g");
                gEnter.append("g").attr("class", "elems");
                gEnter.append("g").attr("class", "x axis");
                gEnter.append("g").attr("class", "y axis");
                
                // Update the outer dimensions.
                svg.attr("width", width)
                    .attr("height", height);

                // Update the inner dimensions.
                var g = svg.select("g")
                    .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

                // Update the individual points
                var nodes = g.select(".elems")
                    .selectAll("rect")
                    .data(data);
                // Update old elements
                nodes
                    .attr("x", function (d) { return xScale(d[0]); })
                    .attr("y", function (d) { return yContentScale(d[1]); })
                    .attr("fill", function (d) { return countScale(d[2]); });
                // Create new elements
                var rects = nodes.enter()
                    .append("rect")
                    .attr("x", function (d) { return xScale(d[0]); })
                    .attr("y", function (d) { return yContentScale(d[1]); })
                    .attr("width", rectSize[0])
                    .attr("height", rectSize[1])
                    .attr("fill", function (d) { return countScale(d[2]); });
                if (tooltipText) {
                    rects.append("title")
                        .html(function (d) { return tooltipText(d) });
                }
                // Remove old elements
                nodes.exit()
                    .remove();
                    
                // Update the x-axis
                var xAxis = g.select(".x.axis")
                    .attr("transform", "translate(0," + contentHeight + ")")
                    .style("font-size", axisFontSizePx + "px")
                    .call(xAxisDef);
                xAxis.selectAll("line")
                    .style("shape-rendering", "crispEdges")
                    .style("stroke", axisColor);
                xAxis.selectAll("text")
                    .style("text-anchor", "start")
                    .style("fill", xAxisTextColor);

                // Update the y-axis
                var yAxis = g.select(".y.axis")
                    .style("font-size", axisFontSizePx + "px")
                    .call(yAxisDef);
                yAxis.selectAll("line")
                    .style("shape-rendering", "crispEdges")
                    .style("stroke", axisColor);
                yAxis.selectAll(".tick text")
                    .attr("x", 2)
                    .style("text-anchor", "start")
                    .style("fill", yAxisTextColor);
                yAxis.selectAll(".tick:last-of-type text")
                    .attr("y", 8);
                yAxis.selectAll(".tick:first-of-type text")
                    .attr("y", -8);
            });
        }

        chart.width = function (_) {
            if (!arguments.length) return width;
            width = _;
            return chart;
        }

        chart.height = function (_) {
            if (!arguments.length) return height;
            height = _;
            return chart;
        }

        chart.rectSize = function (_) {
            if (!arguments.length) return rectSize;
            rectSize = _;
            return chart;
        }

        chart.x = function (_) {
            if (!arguments.length) return xAccessor;
            xAccessor = _;
            return chart;
        }

        chart.y = function (_) {
            if (!arguments.length) return yAccessor;
            yAccessor = _;
            return chart;
        }

        chart.count = function (_) {
            if (!arguments.length) return countAccessor;
            countAccessor = _;
            return chart;
        }

        chart.xFormat = function (_) {
            if (!arguments.length) return xFormat;
            xFormat = _;
            return chart;
        }

        chart.yFormat = function (_) {
            if (!arguments.length) return yFormat;
            yFormat = _;
            return chart;
        }

        chart.colorRange = function (_) {
            if (!arguments.length) return colorRange;
            colorRange = _;
            return chart;
        }

        chart.tooltipText = function (_) {
            if (!arguments.length) return tooltipText;
            tooltipText = _;
            return chart;
        }

        return chart;
    };
})();

src/latencyHeatmap.js
import {extent} from "d3-array";
import {axisBottom, axisLeft} from "d3-axis";
import {nest} from "d3-collection";
import {rgb} from "d3-color";
import {interpolateRgb} from "d3-interpolate";
import {scaleLinear, scaleTime} from "d3-scale";

export default function() {
    var margin = { top: 1, right: 1, bottom: 11, left: 1 },
    width = 600,
    height = 400,
    rectSize = undefined,
    xAccessor = function (d) { return d[0]; },
    xFormat = undefined,
    yAccessor = function (d) { return d[1]; },
    yFormat = undefined,
    countAccessor = function (d) { return d[2]; },
    tooltipText = undefined,
    colorRange = [rgb('#FFFFFF'), rgb('#F03524')],
    axisColor = "#ddd",
    xAxisTextColor = "#999",
    yAxisTextColor = "#000",
    axisFontSizePx = 11;

    function chart(selection) {
        selection.each(function (data) {
            // Convert data to standard representation
            data = data.map(function (d, i) {
                return [xAccessor.call(data, d, i), yAccessor.call(data, d, i), countAccessor.call(data, d, i)];
            });

            // If rectSize is undefined, its dynamically calculated based
            // on width & height.  Otherwise, width & height are dynamically
            // calculated based on rectSize.
            var nCols = nest()
                .key(function (d) { return d[0] })
                .entries(data)
                .length;
            var nRows = nest()
                .key(function (d) { return d[1] })
                .entries(data)
                .length;
            if (rectSize) {
                width = rectSize[0] * nCols + margin.left + margin.right;
                height = rectSize[1] * nRows + margin.top + margin.bottom;
            } else {
                rectSize = [(width - margin.left - margin.right) / nCols, (height - margin.top - margin.bottom) / nRows];
            }

            var contentWidth = width - margin.left - margin.right,
                contentHeight = height - margin.top - margin.bottom;

            // Calculate all the extents once
            var xExtent = extent(data, function (d) { return d[0]; });
            var yExtent = extent(data, function (d) { return d[1]; });
            var countExtent = extent(data, function (d) { return d[2]; });

            var xScale = scaleTime()
                .domain(xExtent)
                .range([0, contentWidth - rectSize[0]]);
            var yAxisScale = scaleLinear()
                .domain(yExtent)
                .range([contentHeight, 0]);
            var yContentScale = scaleLinear()
                .domain(yExtent)
                .range([contentHeight - rectSize[1], 0]);
            var countScale = scaleLinear()
                .domain(countExtent)
                .interpolate(interpolateRgb)
                .range(colorRange);

            // Select the svg element, if it exists.
            var svg = selection
                .selectAll("svg")
                .data([data]);
            // Otherwise, create the skeletal chart.
            var g = svg.enter()
                .append("svg")
                .append("g");
            g.append("g").attr("class", "elems");
            g.append("g").attr("class", "x axis");
            g.append("g").attr("class", "y axis");
            svg = selection.selectAll("svg");
            
            // Update the outer dimensions.
            svg.attr("width", width)
                .attr("height", height);

            // Draw the rectangles
            var gElems = svg.select('.elems')
                .attr('class', 'elems')
                .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
            var update = gElems
                .selectAll("rect")
                .data(data);
            var enter = update.enter();
            var exit = update.exit();
            // Remove old elements not present in new data
            exit.remove();
            // Update old elements
            update
                .attr("x", function (d) { return xScale(d[0]); })
                .attr("y", function (d) { return yContentScale(d[1]); })
                .attr("fill", function (d) { return countScale(d[2]); });
            if (tooltipText) {
                update
                    .selectAll('title')
                    .html(function (d) { return tooltipText(d); });
            }
            // Create new elements
            var newRects = enter
                .append("rect")
                .attr("x", function (d) { return xScale(d[0]); })
                .attr("y", function (d) { return yContentScale(d[1]); })
                .attr("width", rectSize[0])
                .attr("height", rectSize[1])
                .attr("fill", function (d) { return countScale(d[2]); });
            if (tooltipText) {
                newRects
                    .append("title")
                    .html(function (d) { return tooltipText(d) });
            }
                
            // Update the x-axis
            var xAxisDef = axisBottom(xScale)
                .tickSizeInner(-contentHeight)
                .tickSizeOuter(0);
            if (xFormat)
                xAxisDef.tickFormat(xFormat);
            var xAxis = g.select(".x.axis")
                .attr("transform", "translate(0," + contentHeight + ")")
                .style("font-size", axisFontSizePx + "px")
                .call(xAxisDef);
            // Remove axis line
            xAxis.selectAll("path").remove();
            // Update color of axis lines
            xAxis.selectAll("line")
                .style("shape-rendering", "crispEdges")
                .style("stroke", axisColor);
            // Update color of axis labels
            xAxis.selectAll("text")
                .style("text-anchor", "start")
                .style("fill", xAxisTextColor);

            // Update the y-axis
            var yAxisDef = axisLeft(yAxisScale)
                .tickSizeInner(-contentWidth)
                .tickSizeOuter(0)
                .ticks(1);
            if (yFormat)
                yAxisDef.tickFormat(yFormat);
            var yAxis = g.select(".y.axis")
                .style("font-size", axisFontSizePx + "px")
                .call(yAxisDef);
            // Remove axis line
            yAxis.selectAll("path").remove();
            // Update color of axis lines
            yAxis.selectAll("line")
                .style("shape-rendering", "crispEdges")
                .style("stroke", axisColor);
            // Update color of axis labels
            yAxis.selectAll(".tick text")
                .attr("x", 2)
                .style("text-anchor", "start")
                .style("fill", yAxisTextColor);
            yAxis.selectAll(".tick:last-of-type text")
                .attr("y", 8);
            yAxis.selectAll(".tick:first-of-type text")
                .attr("y", -8);
        });
    }

    chart.width = function (_) {
        if (!arguments.length) return width;
        width = _;
        return chart;
    }

    chart.height = function (_) {
        if (!arguments.length) return height;
        height = _;
        return chart;
    }

    chart.rectSize = function (_) {
        if (!arguments.length) return rectSize;
        rectSize = _;
        return chart;
    }

    chart.x = function (_) {
        if (!arguments.length) return xAccessor;
        xAccessor = _;
        return chart;
    }

    chart.y = function (_) {
        if (!arguments.length) return yAccessor;
        yAccessor = _;
        return chart;
    }

    chart.count = function (_) {
        if (!arguments.length) return countAccessor;
        countAccessor = _;
        return chart;
    }

    chart.xFormat = function (_) {
        if (!arguments.length) return xFormat;
        xFormat = _;
        return chart;
    }

    chart.yFormat = function (_) {
        if (!arguments.length) return yFormat;
        yFormat = _;
        return chart;
    }

    chart.colorRange = function (_) {
        if (!arguments.length) return colorRange;
        colorRange = _;
        return chart;
    }

    chart.tooltipText = function (_) {
        if (!arguments.length) return tooltipText;
        tooltipText = _;
        return chart;
    }

    return chart;
}